sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -0,0 +1,293 @@
+        s" max size: ${rightStats.bytesByPartitionId.max}"
+      logDebug(
+        s"""
+          |Try to optimize skewed join.
+          |Left side partition size: $leftSizeInfo
+          |Right side partition size: $rightSizeInfo
+        """.stripMargin)
+
+      val skewedPartitions = mutable.HashSet[Int]()
+      val subJoins = mutable.ArrayBuffer[SparkPlan]()
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val leftMapIdStartIndices = if (isLeftSkew && supportSplitOnLeftPartition(joinType)) {
+          getMapStartIndices(left, partitionId)
+        } else {
+          Array(0)
+        }
+        val rightMapIdStartIndices = if (isRightSkew && supportSplitOnRightPartition(joinType)) {
+          getMapStartIndices(right, partitionId)
+        } else {
+          Array(0)
+        }
+
+        if (leftMapIdStartIndices.length > 1 || rightMapIdStartIndices.length > 1) {
+          skewedPartitions += partitionId
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getNumMappers(left)
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getNumMappers(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO: we may can optimize the sort merge join to broad cast join after
+            //       obtaining the raw data size of per partition,
+            val leftSkewedReader = SkewedPartitionReaderExec(
+              left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+            val rightSkewedReader = SkewedPartitionReaderExec(right, partitionId,
+              rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              s1.copy(child = leftSkewedReader), s2.copy(child = rightSkewedReader))
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.nonEmpty) {
+        val optimizedSmj = smj.transformDown {
+          case sort @ SortExec(_, _, shuffleStage: ShuffleQueryStageExec, _) =>
+            sort.copy(child = PartialShuffleReaderExec(shuffleStage, skewedPartitions.toSet))
+        }
+        subJoins += optimizedSmj
+        UnionExec(subJoins)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.ADAPTIVE_EXECUTION_SKEWED_JOIN_ENABLED)) {
+      return plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
```

Comment:
We could loose an optimization opportunity here. You could have a situation where we converted a SMJ->BHJ and where there is a shuffled join on top of this. Anyway this will require some more changes.

Reply:
This might even be a correctness problem if the LocalShuffleReader produces a partitioning that is actually leveraged later in the stage.

Reply:
We can remove these "reader" cases here:
1. this rule applies first, so we won't see these readers anyway.
2. even if this rule is not the first to apply, we should not skip them otherwise we could miscount. we are not matching them in transform anyway, so we are safe to ignore them here.